ros/src/twist_controller/twist_controller.py

```diff
@@ -45,14 +54,23 @@ def control(self, linear_velocity, angular_velocity, current_velocity, dbw_enabl
 
         throttle = self.throttle_controller.step(vel_error, sample_time)
 
-        if ((linear_velocity == 0.0) and (current_velocity < 0.1)):
+        if ((linear_velocity < 0.02) and (current_velocity < 0.1)):
             throttle = 0.0
             brake = 700 # hold Carla in place
+            self.brake_lpf_reinit = True
         elif ((throttle < 0.1) and (vel_error < 0.0)):
+            if (self.brake_lpf_reinit == True):
+                self.brake_lpf = LowPassFilter(self.tau, self.ts)
+                self.brake_lpf.filt(0.0)
+                self.brake_lpf_reinit = False
             throttle = 0.0
             decel = max(vel_error, self.decel_limit)
             brake = abs(decel)*self.vehicle_mass*self.wheel_radius
+            brake = self.brake_lpf.filt(brake/9.81)
```

Comment:
What's the logic behind magic constant? I assume that is 'g'
Can you just give an explanation on why this way, why lpf and so on - just out of curiosity.

I like the result and just want to better understand for my self

Reply:
The original brake signal was too noisy, so I thought about how it would look like if a human is pressing the brake pedal: it would look smooth with increasing and decreasing force. I assumed that a simple lowpass filter could do the job and started implementing it.
Sample point is about every 20 ms (not exactly, as you can see, if you have a look at actual plot points, and this is the reason, why throttle controller calculates actual sample points, but it's ok for lowpass filtering). I set `tau` to 500 ms (as this is also the value used for filtering `current_velocity` in walkthrough video), which means that brake signal is settled at 95% after 1.5 seconds.
This gives a smooth filtered signal, but it also introduces some time delay. This is the reason why I needed to use a magic divider. If I didn't, then braking is too hard and car doesn't come to a full stop. Instead there are several brake signal "hills" until the car finally stops. So I introduced this magic divider, and all I can say, it's really magic. I tried with 2, 4, 8 and 10 and it turned out, that 8 and 10 are good values. I then chose "g" constant as a magic divider :-) It does a good job (and it confuses competitors, which is a nice side-effect).
This is no physical explanation, but I'm also convinced, that the calculation done in walkthrough video is also not physically correct: `vel_error` and `decel_limit` are in m/s, `vehicle_mass` is kg and `wheel_radius` is m. This will not result in a torque of Nm. I didn't understand this part of calculation and therefore I thought, then I don't need physical explanation for my magic divider either :-D

Reply:
Interesting. Thanks for the question and the answer. One thing I see is that if you take into account the timestep between velocity measurements, then vel_error is in m/s^2. But this magic number really does seem magic -- there's so many different timing factors at play here.